Add "pretty printing"

jmuchovej opened the pull request:
The POMDP file spec supports enum-like representation of `states`/`actions`/`observations`.

Would this be an acceptable PR? (I did it for myself, but figured I'd add it upstream if y'all wanted it.)

Essentially, there's a "pretty printing" mode that uses the string representation of each `states`/`actions`/`observations` and uses those to generate a POMDP file. It does take the longer-form setup (1 line per pairing, vs. the matrix setup), but it's straightforward to support the matrices in the "pretty printing" mode.

I did this before #10 was accepted, so I'll need to rebase, but figured I'd open the PR before rebasing.

jmuchovej committed:
Updating to POMDPTools

POMDPModelTools was renamed to POMDPTools. To ensure future compatibility, POMDPFiles needs to use POMDPTools.

jmuchovej committed:
Add pretty printing state/action/obs names

States, actions, and observations are currently only printed numerically (thus making it difficult to know the corresponding state / action / observation).

This adds the ability to (naïvely) pretty-print the "names" (via `Base.string`) of the state / action / observation.

mykelk commented:
Any opinions on this @zsunberg ? Does it make sense for it to be in `POMDPFiles`? Or somewhere else?

zsunberg commented:
I think it is a good idea to have here since it is a valid POMDP file format according to
https://pomdp.org/code/pomdp-file-spec.html

mykelk commented:
Cool. Let's do it.

zsunberg commented:
@jmuchovej make sure to mark this PR as not a work in progress when it is ready.

In the long term, having a keyword argument for `Base.write(::IO, ::POMDP)` is definitely not the right way to give users access to the option. In fact, I don't think `Base.write` is appropriate for this at all (#14 ), but since that would require a refactor with some design decisions, I'm happy to accept this PR with the keyword argument.

Ideally, someday, we would also be able to read files written with state names in and preserve the names in the POMDP representation, but that is probably for another day.

jmuchovej commented:
Ooh. I see. 

Aside: it seems that `POMDPFiles` exports `POMDPFile` (which is presumably a struct), but I couldn't find a definition for it?

If it were [re]defined, then perhaps that could be used instead of a keyword argument? (Though this seems unwieldy, it could something like...
```julia
Base.write(io::IO, pomdp::POMDP) = ...
Base.write(io::IO, pomdpfile::POMDPFile) = ...
```
So then it retains backwards compatibility, but sidesteps the need for the keyword argument.)

Additionally, I just did some stuff with [EzXML][EzXML.jl] and they export a `prettyprint([io::IO], ...)` function. So perhaps, that could be a suitable alternative to in line with your suggestion in #14, maybe `Base.print` is for the native (numeric) representation and `prettyprint` would be for this representation?

[EzXML.jl]: https://github.com/JuliaIO/EzXML.jl

jmuchovej committed:
Merge branch 'master' into add-pretty-printing

zsunberg commented:
I don't think we need backwards compatibility since only one or two packages depend on this and semver makes it so that upgrading isn't terrible.

I actually think that we should just copy `read_pomdp` and have a function called `write_pomdp` rather than `Base.write`. That can have keyword arguments. There is no need to be fancy.

zsunberg commented:
Note that the CI tests are not passing, please take a look and see if you can fix it.

jmuchovej commented:
So, the CI is failing because #10 introduced `POMDPModels` as a direct dependency, rather than a test dependency. (I removed it as a direct dependency to more closely match the original repo I cloned, pre #10.)

I know how to fix it, but I'm not sure where to put `POMDPModels`.

jmuchovej committed:
Re-add `POMDPModels` to `extras`

zsunberg commented:
POMDPModels is now a direct dependency because the package uses TabularPOMDP.

We should use `using POMDPModels: TabularPOMDP` in the main module file to make this clear.

jmuchovej committed:
Move POMDPModels to deps; spec in module file

jmuchovej commented:
Oh! I see now. Completely missed that in `src/read.jl`. 😅 I moved it to the main module file.

zsunberg commented on src/write.jl:
```diff
@@ -117,3 +145,36 @@ function Base.write(io::IO, pomdp::POMDP)
 	end
 end
+
+function pretty_print(io::IO, pomdp::POMDP)
+	_states       = ordered_states( pomdp)
+	_actions      = ordered_actions(pomdp)
+	_observations = ordered_observations(pomdp)
+
+	println("# ------------------------------------------------------------------------")
+	println("# TRANSITIONS")
+	println("T: * : * : * 0.0")
+	for action=_actions, state=_states, statep=_states
+		T = transition(pomdp, state, action)
+		println(io, "T: $(action) : $(state) : $(statep) : $(pdf(T, statep))")
+	end
+	println(io)
+
+	println("# ------------------------------------------------------------------------")
+	println("# OBSERVATIONS")
+	println("O: * : * : * 0.0")
+	for action=_actions, statep=_states, obs=_observations
+		O = observation(pomdp, action, statep)
+		println(io, "O: $(action) : $(statep) : $(obs) $(pdf(O, obs))")
+	end
+	println(io)
+
+	println("# ------------------------------------------------------------------------")
+	println("# REWARDS")
+	println("R: * : * : * : * 0.0")
+	for action=_actions, state=_states, statep=_states, obs=_observations
+		r = reward(pomdp, state, action, statep, obs)
+		println(io, "R: $(action) : $(state) : $(statep) : $(obs) $(r)")
+	end
+	println(io)
+end
```
Ah, I did not realize that this function was in this PR. It should not be here unless it has something to do with the POMDP file specification https://pomdp.org/code/pomdp-file-spec.html

zsunberg replied:
If you would like to contribute this, it should go in POMDPTools. However, it should also have a more specific name since there are a lot of ways that one might want to "pretty print" a POMDP. I suggest something like `print_distribution_tables`.

jmuchovej replied:
Hm. I'm not sure I understand what you mean here.

> It should not be here unless it has something to do with the POMDP file specification

Are you referring to lines 154-155, 163-164, and 172-173? While these aren't in the spec, they are valid comments. Perhaps it's the sections are self-evident because they're grouped by `T, O, R` anyways. (So I'm happy to remove it.)

> If you would like to contribute this, it should go in POMDPTools. ...

Perhaps this is why you marked it resolved, but this format is explicitly part of the POMDP spec. The idea behind this formatting is to avoid:
1. Numeric representations.
2. Matrices that are semi-opaque in their representation (because you have to keep in mental memory, or annotate, what columns and rows map on to).

e.g., I work with some folks that are more familiar with the `.pomdp` representation for POMDPs and have never used `POMDPs.jl` (so it's useful for them to have this kind of cleaned representation in the `.pomdp` format since that's the framing they understand).

zsunberg replied:
Yeah, I resolved this because I realized that it is part of the spec! Sorry for the confusion!

zsunberg commented on src/write.jl:
```diff
@@ -26,17 +26,45 @@ end
+	end
+
+	n_actions = length(as)
+	if pretty && n_actions < 20
```
The silent `n_actions < 20` check is not good practice. The user should be allowed to do it with 21 actions if they want. At minimum, if an automatic "safety" feature like this is added it should at least notify the user with a message if it happens.

jmuchovej replied:
Agreed. I wasn't sure how to implement this. Do you think this ought to go in the `POMDPFile` struct or something similar?

zsunberg replied:
I would recommend just letting people do it with however many they want. They are the ones who will have to suffer the consequences of extremely long lines in their pomdp files!

zsunberg replied:
I don't see any side effects for anyone else if they want to do it.

jmuchovej replied:
I've opted to remove this. Perhaps I can leave a TODO to implement some kinda `POMDPFile` struct that contains this data, but in retrospect, people can pick their poison. 😅

zsunberg commented on src/write.jl:
```diff
@@ -26,17 +26,45 @@ end
+
+	n_states = length(ss)
+	if pretty && n_states < 20
+    	println(io, "states: ", join([string(s) for s=ss], " "))
```
We should not use just `string` for this. `string` will produce things like `State([1, 2])` which will be extremely difficult for a parser to handle. I think we should by default replace all non-alphanumeric characters in the output of `string` with underscores and allow the user to override this.

jmuchovej replied:
I did `string(s)` to allow users to implement their own `string` method for the state. I'm not super familiar with Julia, so it's possible another string-like method would be suitable, but I figure that "pretty printing" does have some overhead on the user's end for what constitute's "readable" names.

zsunberg replied:
The downside of having them specialize `string` for this specific use is that `string` will behave that way everywhere else in julia. It may well be that the easiest-to-read representation of a state has spaces in it. But in this context, the representation *cannot* have spaces. So if they override string just for this purpose it will muck up everything else.

There are two alternatives
1. Have a function like `POMDPFiles.state_string` that the user can specialize for their type if they want.
2. Just let the user use any function that they pass in as an argument.

I prefer 2 because it is more flexible and easier to change later if they want. It would look something like this:
```julia
function pretty_print(io::IO, m::POMDP; name::Function=x -> non_alphanumeric_to_underscore(string(x)))
    # ...
    println(io, "states: ", join([name(s) for s=ss], " "))
    # ...
end
```
then you write `non_alphanumeric_to_underscore` using regexes or something and put it in this package. Then there is a sensible default, and if the user wants to override it, they can do
```julia
pretty_print(io, m, name = s -> "$(s.x)_$(s.y)")
```
or whatever they want.

jmuchovej replied:
I went the route of individual functions for `states`, `actions`, and `observations`.

zsunberg reviewed (changes requested):
Thanks for the contribution!

This will need a bit more work before merging. See if you can fix the comments below. Then we can talk a bit more about what to name the functions.

jmuchovej committed:
rm arbitrary cutoff; give prettyprint lambdas

There was a 20 `state`/`action`/`observation` cut-off for pretty printing (arbitrarily chosen). Instead, we should let people pick their poison.

Additionally, we forced the use of `string`, but this should really be individual lambdas for each `state`/`action`/`observation` with a default that just enforces ASCII and strips repeating `_`.

jmuchovej commented:
I did some premature refactoring. `print` contains the original implementation that opts for a numeric representation, `prettyprint` is self-explanatory (I hope 😅).

They aren't exported at the moment, but I imagine they probably need to be.

jmuchovej committed:
Reformat exports

Rewrite exports so that `read_pomdp` is visible and remove `POMDPFile` export which is never defined anyway.

jmuchovej committed:
Add link to POMDP spec on pomdps.org

jmuchovej committed:
Fixing typo in prettyprint/obs

Referred to current observation by old var `obs` instead of the newer terse one `o`.

zsunberg reviewed (commented):
@jmuchovej , thanks again for your contribution! This looks great! Would you be willing to make a test for this? Perhaps along the lines of
```julia
struct TestPOMDP <: POMDP{Int, Int, Int} end
POMDPs.states(::TestPOMDP) = (1,)
POMDPs.transition(::TestPOMDP) = Deterministic(1)
# etc.

@test sprint(POMDPFiles.write, TestPOMDP()) == """
<file contents>
"""
```
If you don't have time for that, no worries - I didn't ask you to do this up front. I will probably accept the PR anyways.

I think for function naming we should follow the convention of CSV.jl and have `POMDPFiles.read` and `POMDPFiles.write` that are differ. If you are willing to make that change, great! If not, I can do it. `POMDPFiles.prettyprint` would be ok too if you want to have a separate function rather than just an option.

zsunberg committed:
Merge branch 'master' into add-pretty-printing

zsunberg commented:
(I just pushed changes turning all the tabs into 4 spaces)

jmuchovej commented:
> I think for function naming we should follow the convention of CSV.jl and have `POMDPFiles.read` and `POMDPFiles.write` that are differ.

Should we follow `CSV.jl` over something like `FileIO.jl`? (I guess maybe not ‘cause `CSV.jl` is more highly starred than any singular `FileIO.jl` package?)

> Would you be willing to make a test for this?

Definitely! I might not get to this until later today or the weekend, though.

Also, if the functions are changing, what other packages need to be updated? (I can look into this when writing tests, I just figured you might know offhand.)

zsunberg commented:
Oh, I did not know about FileIO! After an initial glance it seems like a good idea. I don't know enough to have a strong vote between FileIO and CSV, but my initial reaction is that FileIO might be better.

jmuchovej commented:
`FileIO.jl` can also let use do things like dispatch on file extensions (which I think is cool just for kicks). Point being, it might be feasible to unify `POMDPXFiles` and `POMDPFiles` since they're fairly closely tied to begin with (though definitely not necessary right now).

jmuchovej commented:
Since we're going to go for the `FileIO` interface, what do you think of skipping the transition _this_ PR and _just_ having it focus on adding "pretty printing" (and associated tests)?

I'm down to the `FileIO` transition, but I think doing that here will be half-hearted 'cause the `FileIO` interface should also apply to all the read-like functions as well.

zsunberg commented:
I like the way you're thinking, @jmuchovej !! I'm fine with accepting this with just the pretty printing and doing FileIO in another PR.

I will merge this once the tests pass.

jmuchovej committed:
Update printing to use `numericprint` and `symbolicprint`

jmuchovej committed:
Refactor tests, add writer tests

jmuchovej commented:
Ok! I think these are all good to go now (tests and all).
I'll start on the `FileIO.jl` bits once this is merged.

Also, I figure, since this is gonna be a short-lived series of functions, maybe there's no need for a patch bump? (Since the next one is gonna be a minor version bump anyway (afaik, Julia considers <1 minor bumps to be breaking).)

zsunberg commented:
Wow @jmuchovej , this has turned into quite a beautiful PR. Thank you for making such a high-quality contribution to this package!

> Also, I figure, since this is gonna be a short-lived series of functions, maybe there's no need for a patch bump? (Since the next one is gonna be a minor version bump anyway (afaik, Julia considers <1 minor bumps to be breaking).)

Yes, I was just going to wait for FileIO to tag a new version. But if it is going to take you a while, then let me know and I can tag a version with just this.
